### ISSUE TITLE
Proposal: Benchmarks

### ISSUE THREAD

**Issue by BrianHicks** [elm-community/elm-test#92]
TL;DR: Let's add benchmarks to elm-test. @Skinney's elm-array-exploration already has an implementation that could be an inspiration for the API. We have examples of prior art where benchmarks have been useful to make software better. I also have some opinions at the bottom! Feedback on any of this is much appreciated.

## Motivating and Demotivating Examples

- [Skinney/elm-array-exploration](https://github.com/Skinney/elm-array-exploration) is about improving performance with a drop-in API. (Please feel free to correct me if that's wrong, @Skinney.) It includes a good place to start on benchmark API discussion.
- [Blazing Fast HTML, Round Two](http://elm-lang.org/blog/blazing-fast-html-round-two) talks about benchmarking. In this case, we're comparing implementations of *frameworks*, not *units* (for lack of more precise terms.) The actual implementation of benchmarks in elm-test should be in terms of units. Why? First, they're more honest (discussed below, and not to say that Evan's post isn't honest. In fact, [he went to great lengths](http://elm-lang.org/blog/blazing-fast-html-round-two#methodology) to avoid the normal benchmark dishonesty.) Second, they're easier to implement and understand, and fall more easily within the boundaries of elm-test's problem space.
- I am curious about the performance of my app implementations, particularly in parsers and decoders. I don't have a clean way to test this in a consistent way, and the iffy ways I *do* have are convenient. Because of this, I don't benchmark my implementations. I'd like to change that!
- I'm also interested if the Set implementation that I'm writing to teach pure FP data structures is actually comparable to core's performance. But again, no affordance.

## Prior Art

- [*Pushing The Limits of Web Browsers … or Why Speed Matters* by Lars Bak](https://www.infoq.com/presentations/Performance-V8-Dart) talks about V8 performance, and mentions benchmarking as important to prevent performance regressions. (link courtesy of @rtfeldman, I haven't seen the talk myself.)
- Go's standard testing library has benchmark implementation built in. Therefore, benchmarks get written and used, and are fairly common in the wild. ([`testing` package documention for this](https://golang.org/pkg/testing/#hdr-Benchmarks))
- [fredcy/elm-benchmark](https://github.com/fredcy/elm-benchmark) implements benchmarking in Elm.

## My Opinions (Hot Take Alert!)

Elm is meant to be fast, so we should provide a way to measure and quantify that. Benchmarks are a useful tool for that.

That said, *dishonest* benchmarks are awful, and can destroy credibility. This can happen with benchmark selection for marketing purposes by selecting benchmarks to publish in which you hit the sweet spot of product A and regressive behavior of product B then generalizing that "A is faster than B" in all cases. Because of this history, we should set a standard that implementations should only be compared to one another and in a consistent environment. Apples to apples.

Also, I think performance regressions are almost as bad as feature regressions. When we add features to libraries and applications, we need to be able to at least see the tradeoffs we're making regarding performance. Benchmarking is one means to this end.

**Comment by Skinney** [comment]
elm-array-exploration isn't really about performance. It's about being stable (core implementation has many bugs) and more understandable, while not sacrificing performance. As it stands, elm-array-exploration is much slower in some cases, and faster in others.

The benchmark "library" used in that repo is a very crude wrapper around benchmark.js. While it is functional and guides the implementation of the array library, it also exploits the fact that it isn't a library to be used by others. Benchmarks can, for instance, return whatever type they want to (Int, String, List, whatever) because it made things simpler. This should probably not possible in an official benchmark library.

I do agree that an official benchmarking library should be available for Elm, but I don't have an opinion on whether it should be part of elm-test, or be its own library.

**Comment by BrianHicks** [comment]
Regarding performance, that makes sense! Thanks for clarifying. I'm really glad that you're considering the performance implications of your work though, for the reasons I outlined above.

What parts of your benchmark implementation would you consider "good"? Those are the things I think we would want to consider for elm-test.

And as to the inclusion in elm-test, I'd say that fuzzing has about as much to do with testing as benchmarks do. They're all ways to verify properties about the systems we're building. That said, it does seem like it might need to be a separate mode of operation! Go does this in their implementation; you have to pass a `-bench` flag to get the benchmarks to run.

**Comment by mgold** [comment]
The benchmarking for array-exploration seems to be closely related to [fredcy/elm-benchmark](https://github.com/fredcy/elm-benchmark). As mentioned, it uses a JS library and therefore it would not be possible to include in elm-test.

Moreover, I disagree with the claim that "fuzzing has about as much to do with testing as benchmarks do", if I may understand "testing" as unit testing. A fuzz test is a way to generate many unit tests, and to talk about the (probable) behavior of a function over all of its inputs instead of the (definite) behavior over select inputs. In both cases, we care about the correctness of output for a given input, not how long it takes.

If someone wants to either create a pure Elm benchmark library, or maintain the native one and make it easier to install, then *they* can add an expectation that a particular function completes in less than a certain amount of time, to catch perf regressions. It would be very possible to do that using the interface we already expose, especially `Expect.pass` and `Expect.fail`. `elm-html-test` is already doing that.

**Comment by zkessin** [comment]
@BrianHicks for me at least FuzzTests are often my primary tests, i will often write them as my only unit tests

**Comment by zkessin** [comment]
FWIW I think performance testing should be a separate library, (Though with a similar api would be good). the issue with performance tests is that they often do not give a specific *YES*/*NO* but something that may need some thinking about as in it took 31ms do to x, is that good or bad?

**Comment by BrianHicks** [comment]
@mgold I didn't see @fredcy's implementation. Added to the Prior Art section above in the top post. Good catch!

You also bring up a good point about the semantic difference between benchmarks and fuzzing. I was referring to the execution mode, not the semantics. There's clearly a difference there, you're right. I could have been more clear.

The semantic difference between benchmarks and unit tests is pretty important. You actually don't want a benchmark to fail if it's too slow, because it will perform differently on different computers. For example, my Go benchmarks run faster in CI than on my Mac because there's less contention. So there's actually not an absolute value of performance, only a relative one.

That said, I do think pass/fail semantics are useful, but only in the relative sense (to @zkessin's point). I think the best physical analogy is a speed limit. In the US, 20/25/35 MPH are common on residential streets and 55/60/75 on the highways. The value of "fast" or "slow" differs depending on where you're driving. Plus, you're more likely to be be pulled over going 60 in a 25 than 35 in a 25. A little bit off is sometimes OK, but a lot off is not. I'm actually not sure how to translate that to code, but I think it's a useful analogy to understand the problem!

Regarding where this code should live: let's figure out what it needs to look like first. This could easily become a bikeshed discussion.

So my question: given that we have @fredcy's implementation (and his README lists more prior art), what do we do with the measurements? Is the speed limit analogy useful? How could we translate that into a model useful to users? Could we?

**Comment by mgold** [comment]
Well, this is leaving the realm of elm-test pretty far behind, but...

It might be worthwhile to define some baseline operation, benchmark that, and define everything else in units of the baseline. Rudimentary tests would be, "I expect this operation to take about 5 baseline units", and the tests fail if it gets noticeably slower. (If a test gets faster, it could mean you've made a legitimate optimization, or it could mean that you're neglecting to do all the required work thereby have introduced a bug. Not sure what to do in this case. Conversely, a test may get slower when you fix a bug.)

The next thing you can do -- and this is very not functional but we're already using native libraries -- is, every time the suite is run to completion, save the results to a file. Then you have some history to go off of to determine whether a result is within the range of expectation.

But sometimes statistics happen, and that will lead to flaky tests. So I'm not sure what we can do here.

**Comment by BrianHicks** [comment]
Keeping history is going to have to be necessary to indicate trends. I wouldn't call it un-functional, though! We keep state in TEA all the time. The difference here is that we need a way to persist and reload it. I don't think that's a huge deal to do.

It's a bigger deal to figure out how to add *meaning* to the data that's produced by benchmarks. I think it would be enough to scope saved state per environment. Differing, for example, between dev and CI (where I would imagine dev would have more samples but more noise.)

As for measurements, I think we can use variables control charts. They're fairly simple (measuring standard deviations from a control point.) This could be the "base units" concept you're thinking of, @mgold? From the [NIST](http://itl.nist.gov/div898/handbook/pmc/section3/pmc32.htm):

```
WECO stands for Western Electric Company Rules 
 
       Any Point Above +3 Sigma  
 ---------------------------------------------    +3 σ LIMIT 
       2 Out of the Last 3 Points Above +2 Sigma  
 ---------------------------------------------    +2 σ LIMIT 
       4 Out of the Last 5 Points Above +1 Sigma  
 ---------------------------------------------    +1 σ LIMIT 
       8 Consecutive Points on This Side of Control Line  
==============================   CENTER LINE 
       8 Consecutive Points on This Side of Control Line  
 ---------------------------------------------    -1 σ LIMIT 
       4 Out of the Last 5 Points Below - 1 Sigma  
----------------------------------------------   -2 σ LIMIT 
       2 Out of the Last 3 Points Below -2 Sigma  
 ---------------------------------------------    -3 σ LIMIT 
       Any Point Below -3 Sigma 

Trend Rules: 6 in a row trending up or down.
             14 in a row alternating up and down.
```

Control point could be decided after a reasonable number of samples (NIST recommends 25 at size 4, minimum.) Before that, we could say "not enough data", use the running median, or ask the user to select a good sample. If a large positive or negative change is detected, we could ask the user to please remove the history that doesn't reflect the new measurement and go from there.

**Comment by mgold** [comment]
It's not a bad idea, but it sounds like something that should be pioneered as a third-party library. I don't have the time or domain knowledge to do it myself.

**Comment by BrianHicks** [comment]
I wasn't asking you to do it. But point taken, and I'll move this discussion to elm-discuss.

**Comment by mgold** [comment]
Okay. Please don't be too upset, I'd love to see this happen! 😄

**Comment by BrianHicks** [comment]
Oh no, not upset at all. :)

**Comment by zkessin** [comment]
I agree, this should be a 3rd party library. though it might be nice if it could be run from the node test runner, that being said it might make more sense to run in a browser

**Comment by BrianHicks** [comment]
Using the existing runners is part of my motivation for suggesting it here. We'd have to reinvent so much, otherwise.

That said, I'm curious of your reasoning why running in the browser might be better?

**Comment by zkessin** [comment]
i figure if I am going to do performance messurement the browser is what matters, that is after all where my users are going to run it, and I don't have a specific reason to assume that performance issues will be the same between node.js and firefox/chrome/ms edge etc

**Comment by zkessin** [comment]
i figure if I am going to do performance messurement the browser is what matters, that is after all where my users are going to run it, and I don't have a specific reason to assume that performance issues will be the same between node.js and firefox/chrome/ms edge etc
